server/src/main/java/org/elasticsearch/index/shard/IndexWriteLoad.java

```diff
@@ -0,0 +1,217 @@
+            shardsWriteLoad.stream().mapToDouble(shardLoad -> shardLoad).toArray(),
+            shardsUptimeInMillis.stream().mapToLong(shardUptime -> shardUptime).toArray()
+        );
+    }
+
+    @Nullable
+    public static IndexWriteLoad fromStats(IndexMetadata indexMetadata, @Nullable IndicesStatsResponse indicesStatsResponse) {
+        if (indicesStatsResponse == null) {
+            return null;
+        }
+
+        final IndexStats indexStats = indicesStatsResponse.getIndex(indexMetadata.getIndex().getName());
+        if (indexStats == null) {
+            return null;
+        }
+
+        final int numberOfShards = indexMetadata.getNumberOfShards();
+        final var indexWriteLoadBuilder = IndexWriteLoad.builder(numberOfShards);
+        final var indexShards = indexStats.getIndexShards();
+        for (IndexShardStats indexShardsStats : indexShards.values()) {
+            final var shardStats = Arrays.stream(indexShardsStats.getShards())
+                .filter(stats -> stats.getShardRouting().primary())
+                .findFirst()
+                // Fallback to a replica if for some reason we couldn't find the primary stats
+                .orElse(indexShardsStats.getAt(0));
+            final var indexingShardStats = shardStats.getStats().getIndexing().getTotal();
+            indexWriteLoadBuilder.withShardWriteLoad(
+                shardStats.getShardRouting().id(),
+                indexingShardStats.getWriteLoad(),
+                indexingShardStats.getTotalActiveTimeInMillis()
+            );
+        }
+        return indexWriteLoadBuilder.build();
+    }
+
+    private final double[] shardWriteLoadLoad;
+    private final long[] shardUptimeInMillis;
+
+    private IndexWriteLoad(double[] shardWriteLoadLoad, long[] shardUptimeInMillis) {
+        assert shardWriteLoadLoad.length == shardUptimeInMillis.length;
+        this.shardWriteLoadLoad = shardWriteLoadLoad;
+        this.shardUptimeInMillis = shardUptimeInMillis;
+    }
+
+    public IndexWriteLoad(StreamInput in) throws IOException {
+        this(in.readDoubleArray(), in.readLongArray());
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeDoubleArray(shardWriteLoadLoad);
+        out.writeLongArray(shardUptimeInMillis);
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.field(SHARDS_WRITE_LOAD_FIELD.getPreferredName(), shardWriteLoadLoad);
+        builder.field(SHARDS_UPTIME_IN_MILLIS.getPreferredName(), shardUptimeInMillis);
+        return builder;
+    }
+
+    public static IndexWriteLoad fromXContent(XContentParser parser) throws IOException {
+        return PARSER.parse(parser, null);
+    }
+
+    @Nullable
+    public Double getWriteLoadForShard(int shardId) {
+        return getWriteLoadForShard(shardId, null);
+    }
+
+    public Double getWriteLoadForShard(int shardId, Double defaultValue) {
```

Comment:
Seems like you think we will need the fallback value versions in a follow-up? All good if so, but if we are not sure yet, I'd prefer to just add that later if needed when we start using it.